Source/JavaScriptCore/wasm/WasmFunctionParser.h

```diff
@@ -1629,6 +1629,54 @@ auto FunctionParser<Context>::parseArrayTypeDefinition(const char* operation, bo
     return { };
 }
 
+// Return true if parsing should continue after, false if parseExpression() should return.
+template <typename Context>
+ALWAYS_INLINE bool FunctionParser<Context>::parseNestedBlocksEagerly()
+{
+    do {
+        int8_t kindByte;
+        BlockSignature inlineSignature;
+
+        // Only attempt to parse the most optimistic case of a single non-ref or void return signature.
+        if (LIKELY(peekInt7(kindByte) && isValidTypeKind(kindByte))) {
+            TypeKind typeKind = static_cast<TypeKind>(kindByte);
+            Type type = { typeKind, TypeDefinition::invalidIndex };
+            if (UNLIKELY(!(type.isVoid() || isValueType(type))))
+                return true;
+            inlineSignature = m_typeInformation.thunkFor(type);
+        } else
+            return true;
+
+        ASSERT(!inlineSignature->argumentCount());
+
+        int64_t oldSize = m_expressionStack.size();
+        Stack newStack;
+        ControlType block;
+        auto result = m_context.addBlock(inlineSignature, m_expressionStack, block, newStack);
+        if (UNLIKELY(!result))
+            return true;
+        ASSERT_UNUSED(oldSize, oldSize - m_expressionStack.size() == inlineSignature->argumentCount());
+        ASSERT(newStack.size() == inlineSignature->argumentCount());
+
+        // Only increment after possible block failures are checked.
+        m_offset++;
+
+        m_controlStack.append({ WTFMove(m_expressionStack), { }, getLocalInitStackHeight(), WTFMove(block) });
```

Comment:
Could this plus the copy below get turned into a little helper?

Reply:
Done